### PR TITLE
Update: pyAMReX v24.03 & v24.04

### PR DESCRIPTION
**Pull request by ax3l**
Update pyAMReX #31915: add latest releases (24.03, 24.04).

### CODE REVIEW

**Comment by ax3l** [comment]
Ready for approval and merge :)

**Comment by ax3l** [comment]
@alalazo or @haampie pls merge :)

File: var/spack/repos/builtin/packages/py-amrex/package.py [hunk 1]
```diff
@@ -10,15 +10,26 @@ class PyAmrex(PythonPackage, CudaPackage, ROCmPackage):
     """AMReX Python Bindings with pybind11"""
 
     homepage = "https://amrex-codes.github.io/amrex/"
+    url = "https://github.com/AMReX-Codes/pyamrex/archive/refs/tags/24.01.tar.gz"
```
Comment (ax3l) [inline var/spack/repos/builtin/packages/py-amrex/package.py]:
```suggestion
    url = "https://github.com/AMReX-Codes/pyamrex/archive/refs/tags/24.03.tar.gz"
```

**Review by WeiqunZhang — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-amrex/package.py [hunk 2]
```diff
@@ -64,15 +76,23 @@ class PyAmrex(PythonPackage, CudaPackage, ROCmPackage):
...
+    depends_on("py-cupy", type="test", when="+cuda")
+
+    tests_src_dir = "tests/"
+
     def setup_build_environment(self, env):
         spec = self.spec
 
         # disable superbuilds: use external dependencies
         env.set("AMREX_INTERNAL", "OFF")
+        env.set("PYAMREX_CCACHE", "ON")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/py-amrex/package.py:90]: Does this need `ccache` as a build dependency?

**Review by tgamblin — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-amrex/package.py (hunk 2 above)
Comment (ax3l) [reply var/spack/repos/builtin/packages/py-amrex/package.py:90]: Oh no, that should be
```suggestion
        env.set("PYAMREX_CCACHE", "OFF")
```

**Comment by kwryankrattiger** [comment]
I was looking at the gitlab CI run to see if we had py-amrex tests running anywhere and I don't see it. I am assuming those are working, but maybe it would be nice to build amrex by itself in a stack? Currently only warpx builds it, but that isn't quite the same.

Verified the checksums.

File: var/spack/repos/builtin/packages/py-amrex/package.py (hunk 2 above)
Comment (ax3l) [reply var/spack/repos/builtin/packages/py-amrex/package.py:90]: #43570

**Comment by ax3l** [comment]
@kwryankrattiger yes, we definitely have to add `py-amrex` now to the stack where we build `amrex`. This is new.

Can you take this on or guide me where I should add it?

**Comment by kwryankrattiger** [comment]
So step 1 is to find a place for it.

I think E4S may be a good place. @eugeneswalker 

If we made warpX depend directly on AMReX vs pulling its own copy that would happen for free without introducing a new root spec. I think I remember that being non-trivial in the past though.

Another odder place could be in the DAV stack as a consumer of insitu tools. It is a bit orthogonal there, but sellable as a downstream tester package for consuming DAV packages.

Another other option is to create a new stack targeting some domain/use cases. This is the most work but possibly worth it if there are a lot missing builds.

**Pull request merged**